In [1]:
import pandas as pd
import numpy as np
import warnings
from scipy import stats
from sqlalchemy import create_engine
warnings.filterwarnings('ignore')
import mysql.connector


In [2]:
conn = mysql.connector.connect(
    host="localhost",
    user = "root",
    password = "ibadat",
    database = "covid_db"
)

In [3]:
#Checking tables in the database
cursor = conn.cursor()
cursor.execute("SHOW TABLES")
tables = cursor.fetchall()

#Printing tables
print("Tables in the database:",tables)

Tables in the database: [('comm_cols_list',), ('covid_monthly_summary',), ('coviddeaths',), ('covidvaccinations',), ('final_covid',)]


In [4]:
cursor.execute("SELECT * FROM coviddeaths LIMIT 50")
result = cursor.fetchall()
result

[('AFG',
  'Asia',
  'Afghanistan',
  '24/02/2020',
  1.0,
  1.0,
  None,
  None,
  None,
  None,
  0.026,
  0.026,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  8.33,
  38928341.0,
  54.422,
  18.6,
  2.581,
  1.337,
  1803.987,
  None,
  597.029,
  9.59,
  None,
  None,
  37.746,
  0.5,
  64.83,
  0.511),
 ('AFG',
  'Asia',
  'Afghanistan',
  '25/02/2020',
  1.0,
  0.0,
  None,
  None,
  None,
  None,
  0.026,
  0.0,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  8.33,
  38928341.0,
  54.422,
  18.6,
  2.581,
  1.337,
  1803.987,
  None,
  597.029,
  9.59,
  None,
  None,
  37.746,
  

In [5]:
#printing columns in deaths table
deaths_cols = [col[0] for col in cursor.description]
print("Total Columns in deaths table:", len(deaths_cols))
print("Columns: ", deaths_cols,"\n")

Total Columns in deaths table: 59
Columns:  ['iso_code', 'continent', 'location', 'date', 'total_cases', 'new_cases', 'new_cases_smoothed', 'total_deaths', 'new_deaths', 'new_deaths_smoothed', 'total_cases_per_million', 'new_cases_per_million', 'new_cases_smoothed_per_million', 'total_deaths_per_million', 'new_deaths_per_million', 'new_deaths_smoothed_per_million', 'reproduction_rate', 'icu_patients', 'icu_patients_per_million', 'hosp_patients', 'hosp_patients_per_million', 'weekly_icu_admissions', 'weekly_icu_admissions_per_million', 'weekly_hosp_admissions', 'weekly_hosp_admissions_per_million', 'new_tests', 'total_tests', 'total_tests_per_thousand', 'new_tests_per_thousand', 'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'new_vaccinations', 'new_vaccinations_smoothed', 'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred', 'people_fully_vaccin

In [6]:
cursor.execute("SELECT * FROM covidvaccinations ")
result_vacc = cursor.fetchall()
result_vacc

[('AFG',
  'Asia',
  'Afghanistan',
  '24/02/2020',
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  8.33,
  54.422,
  18.6,
  2.581,
  1.337,
  1803.987,
  None,
  597.029,
  9.59,
  None,
  None,
  37.746,
  0.5,
  64.83,
  0.511),
 ('AFG',
  'Asia',
  'Afghanistan',
  '25/02/2020',
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  8.33,
  54.422,
  18.6,
  2.581,
  1.337,
  1803.987,
  None,
  597.029,
  9.59,
  None,
  None,
  37.746,
  0.5,
  64.83,
  0.511),
 ('AFG',
  'Asia',
  'Afghanistan',
  '26/02/2020',
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  None,
  8.33,
  54.422,
  18.6,
  2.581,
  1.337,
  1803.987,
  None,
  597.029,
  9.59,
  None,
  None,
  37.746,
  0.5,
  64.83,
  0.511),
 ('AFG'

In [7]:
vaccines_cols = [col[0] for col in cursor.description]
print("Total Columns in vaccinations table:", len(vaccines_cols))  
print("Columns: ", vaccines_cols)

Total Columns in vaccinations table: 37
Columns:  ['iso_code', 'continent', 'location', 'date', 'new_tests', 'total_tests', 'total_tests_per_thousand', 'new_tests_per_thousand', 'new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'positive_rate', 'tests_per_case', 'tests_units', 'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'new_vaccinations', 'new_vaccinations_smoothed', 'total_vaccinations_per_hundred', 'people_vaccinated_per_hundred', 'people_fully_vaccinated_per_hundred', 'new_vaccinations_smoothed_per_million', 'stringency_index', 'population_density', 'median_age', 'aged_65_older', 'aged_70_older', 'gdp_per_capita', 'extreme_poverty', 'cardiovasc_death_rate', 'diabetes_prevalence', 'female_smokers', 'male_smokers', 'handwashing_facilities', 'hospital_beds_per_thousand', 'life_expectancy', 'human_development_index']


In [8]:
#NOW founding the same columns from both tables
common_cols = set(deaths_cols).intersection(set(vaccines_cols))

d1 = set(deaths_cols) - set(vaccines_cols)
print("\nColumns in deaths table but not in vaccinations table:", len(d1))

v1 = set(vaccines_cols) - set(deaths_cols)
print("\nColumns in vaccinations table but not in deaths table:", len(v1))

print("Printing common columns from both tables:    ")
print("*"*50)
print("Common Columns in both tables:", len(common_cols))
for col in common_cols:
    print(col)




Columns in deaths table but not in vaccinations table: 22

Columns in vaccinations table but not in deaths table: 0
Printing common columns from both tables:    
**************************************************
Common Columns in both tables: 37
total_vaccinations
date
new_tests_per_thousand
female_smokers
people_vaccinated_per_hundred
new_vaccinations_smoothed
handwashing_facilities
tests_units
extreme_poverty
gdp_per_capita
new_vaccinations
diabetes_prevalence
continent
new_tests_smoothed
location
people_vaccinated
iso_code
tests_per_case
new_tests
people_fully_vaccinated
total_tests_per_thousand
male_smokers
total_tests
new_tests_smoothed_per_thousand
total_vaccinations_per_hundred
stringency_index
life_expectancy
population_density
new_vaccinations_smoothed_per_million
aged_65_older
cardiovasc_death_rate
positive_rate
median_age
people_fully_vaccinated_per_hundred
human_development_index
aged_70_older
hospital_beds_per_thousand


## **Combining the both tables basis on these columns**

**total Columns = 59**

**Common Columns b/w two tabels = 37**

Given Above ........

**we need also other columns for our bussiness analysis**

Additional Columns : 
- Disease spread patterns
- Healthcare System capacity
- Vaccination effectiveness
- Socioeconomic impact
- Regional disparities
- Time-based trend

In [9]:
query = """
-- Monthly Aggregated COVID-19 Analysis by Country
SELECT
    -- Geographic Identifiers
    d.iso_code AS country_code,
    d.continent AS continent_name,
    d.location AS country_name,
    DATE_FORMAT(STR_TO_DATE(d.date, '%d/%m/%Y'), '%Y-%m-01') AS month_start_date,
    YEAR(STR_TO_DATE(d.date, '%d/%m/%Y')) AS year,
    MONTH(STR_TO_DATE(d.date, '%d/%m/%Y')) AS month,
   
    -- Demographics (constant per country - take first value)
    MAX(d.population) AS total_population,
    MAX(d.population_density) AS population_per_sq_km,
    MAX(d.median_age) AS median_age_years,
    MAX(d.aged_65_older) AS percent_aged_65_plus,
    MAX(d.aged_70_older) AS percent_aged_70_plus,
   
    -- Economic Indicators (constant - take first value)
    MAX(d.gdp_per_capita) AS gdp_per_capita_usd,
    MAX(d.extreme_poverty) AS percent_in_extreme_poverty,
   
    -- Health Infrastructure (constant - take first value)
    MAX(d.cardiovasc_death_rate) AS cardiovascular_death_rate,
    MAX(d.diabetes_prevalence) AS diabetes_prevalence_percent,
    MAX(d.female_smokers) AS female_smoking_rate,
    MAX(d.male_smokers) AS male_smoking_rate,
    MAX(d.handwashing_facilities) AS handwashing_facilities_percent,
    MAX(d.hospital_beds_per_thousand) AS hospital_beds_per_1000_pop,
    MAX(d.life_expectancy) AS life_expectancy_years,
    MAX(d.human_development_index) AS human_development_index_score,
   
    -- COVID Cases Metrics - Monthly Aggregations
    MAX(d.total_cases) AS total_confirmed_cases_month_end,
    SUM(d.new_cases) AS monthly_new_cases,
    AVG(d.new_cases_smoothed) AS avg_daily_cases_7day,
    MAX(d.total_cases_per_million) AS cases_per_million_month_end,
    AVG(d.new_cases_per_million) AS avg_daily_cases_per_million,
    AVG(d.new_cases_smoothed_per_million) AS avg_daily_cases_per_million_7day,
   
    -- COVID Deaths Metrics - Monthly Aggregations
    MAX(d.total_deaths) AS total_confirmed_deaths_month_end,
    SUM(d.new_deaths) AS monthly_new_deaths,
    AVG(d.new_deaths_smoothed) AS avg_daily_deaths_7day,
    MAX(d.total_deaths_per_million) AS deaths_per_million_month_end,
    AVG(d.new_deaths_per_million) AS avg_daily_deaths_per_million,
    AVG(d.new_deaths_smoothed_per_million) AS avg_daily_deaths_per_million_7day,
   
    -- Case Fatality Rate (calculated metric)
    CASE 
        WHEN MAX(d.total_cases) > 0 
        THEN (MAX(d.total_deaths) / MAX(d.total_cases)) * 100 
        ELSE NULL 
    END AS case_fatality_rate_percent,
   
    -- Epidemiological Indicators
    AVG(d.reproduction_rate) AS avg_virus_reproduction_rate,
    MAX(d.reproduction_rate) AS max_virus_reproduction_rate,
   
    -- Healthcare System Load - Monthly Aggregations
    AVG(d.icu_patients) AS avg_icu_patients,
    MAX(d.icu_patients) AS peak_icu_patients,
    AVG(d.icu_patients_per_million) AS avg_icu_patients_per_million,
    MAX(d.icu_patients_per_million) AS peak_icu_patients_per_million,
    AVG(d.hosp_patients) AS avg_hospital_patients,
    MAX(d.hosp_patients) AS peak_hospital_patients,
    AVG(d.hosp_patients_per_million) AS avg_hospital_patients_per_million,
    MAX(d.hosp_patients_per_million) AS peak_hospital_patients_per_million,
    SUM(d.weekly_icu_admissions) AS total_monthly_icu_admissions,
    SUM(d.weekly_hosp_admissions) AS total_monthly_hospital_admissions,
   
    -- Testing Metrics - Monthly Aggregations
    SUM(v.new_tests) AS total_monthly_tests_conducted,
    MAX(v.total_tests) AS total_tests_conducted_month_end,
    MAX(v.total_tests_per_thousand) AS tests_per_thousand_month_end,
    AVG(v.new_tests_per_thousand) AS avg_daily_tests_per_thousand,
    AVG(v.new_tests_smoothed) AS avg_daily_tests_7day,
    AVG(v.positive_rate) AS avg_test_positivity_rate,
    MIN(v.positive_rate) AS min_test_positivity_rate,
    MAX(v.positive_rate) AS max_test_positivity_rate,
    AVG(v.tests_per_case) AS avg_tests_required_per_case,
   
    -- Vaccination Metrics - Monthly Aggregations
    MAX(v.total_vaccinations) AS total_vaccine_doses_month_end,
    MAX(v.people_vaccinated) AS people_with_at_least_one_dose_month_end,
    MAX(v.people_fully_vaccinated) AS people_fully_vaccinated_month_end,
    SUM(v.new_vaccinations) AS monthly_vaccinations,
    AVG(v.new_vaccinations_smoothed) AS avg_daily_vaccinations_7day,
    MAX(v.total_vaccinations_per_hundred) AS vaccine_doses_per_100_month_end,
    MAX(v.people_vaccinated_per_hundred) AS people_vaccinated_per_100_month_end,
    MAX(v.people_fully_vaccinated_per_hundred) AS people_fully_vaccinated_per_100_month_end,
   
    -- Policy Response
    AVG(v.stringency_index) AS avg_government_response_stringency,
    MAX(v.stringency_index) AS max_government_response_stringency,
    MIN(v.stringency_index) AS min_government_response_stringency,
    
    -- Record Count
    COUNT(*) AS days_in_month
FROM
    coviddeaths d
    INNER JOIN covidvaccinations v
        ON d.location = v.location AND d.date = v.date
WHERE
    d.continent IS NOT NULL
GROUP BY
    d.iso_code,
    d.continent,
    d.location,
    DATE_FORMAT(STR_TO_DATE(d.date, '%d/%m/%Y'), '%Y-%m-01'),
    YEAR(STR_TO_DATE(d.date, '%d/%m/%Y')),
    MONTH(STR_TO_DATE(d.date, '%d/%m/%Y'))
ORDER BY
    d.location,
    year,
    month
"""



In [10]:
final_covid = pd.read_sql(query,conn)

In [11]:
final_covid.describe().T

count          mean           std  \
year                                       2793.0  2.020295e+03  4.562957e-01   
month                                      2793.0  5.839241e+00  3.420234e+00   
total_population                           2789.0  4.258748e+07  1.549574e+08   
population_per_sq_km                       2712.0  3.835090e+02  1.778115e+03   
median_age_years                           2618.0  3.059003e+01  9.185053e+00   
...                                           ...           ...           ...   
people_fully_vaccinated_per_100_month_end   325.0  7.013785e+00  1.193142e+01   
avg_government_response_stringency         2523.0  5.761138e+01  2.121929e+01   
max_government_response_stringency         2523.0  6.277997e+01  2.214289e+01   
min_government_response_stringency         2523.0  5.214055e+01  2.295113e+01   
days_in_month                              2793.0  2.902256e+01  5.323341e+00   

                                                min          25%  \
year                                       2020.000     2020.000   
month                                         1.000        3.000   
total_population                            809.000  2142252.000   
population_per_sq_km                          0.137       36.253   
median_age_years                             15.100       22.000   
...                                             ...          ...   
people_fully_vaccinated_per_100_month_end     0.000        0.670   
avg_government_response_stringency            0.000       44.382   
max_government_response_stringency            0.000       49.540   
min_government_response_stringency            0.000       36.340   
days_in_month                                 1.000       30.000   

                                                    50%           75%  \
year                                       2.020000e+03  2.021000e+03   
month                                      5.000000e+00  9.000000e+00   
total_population                           9.449321e+06  3.107294e+07   
population_per_sq_km                       8.430400e+01  2.095880e+02   
median_age_years                           2.970000e+01  3.910000e+01   
...                                                 ...           ...   
people_fully_vaccinated_per_100_month_end  2.610000e+00  8.110000e+00   
avg_government_response_stringency         5.981290e+01  7.355394e+01   
max_government_response_stringency         6.667000e+01  7.963000e+01   
min_government_response_stringency         5.463000e+01  7.037000e+01   
days_in_month                              3.100000e+01  3.100000e+01   

                                                    max  
year                                       2.021000e+03  
month                                      1.200000e+01  
total_population                           1.439324e+09  
population_per_sq_km                       2.054677e+04  
median_age_years                           4.820000e+01  
...                                                 ...  
people_fully_vaccinated_per_100_month_end  9.747000e+01  
avg_government_response_stringency         1.000000e+02  
max_government_response_stringency         1.000000e+02  
min_government_response_stringency         1.000000e+02  
days_in_month                              3.100000e+01  

[63 rows x 8 columns]

In [12]:
final_covid.describe(include = 'all').T

count unique         top freq  \
country_code                                 2793    210         AUS   16   
continent_name                               2793      6      Africa  760   
country_name                                 2793    210   Australia   16   
month_start_date                             2793     16  2021-04-01  210   
year                                       2793.0    NaN         NaN  NaN   
...                                           ...    ...         ...  ...   
people_fully_vaccinated_per_100_month_end   325.0    NaN         NaN  NaN   
avg_government_response_stringency         2523.0    NaN         NaN  NaN   
max_government_response_stringency         2523.0    NaN         NaN  NaN   
min_government_response_stringency         2523.0    NaN         NaN  NaN   
days_in_month                              2793.0    NaN         NaN  NaN   

                                                  mean        std     min  \
country_code                                       NaN        NaN     NaN   
continent_name                                     NaN        NaN     NaN   
country_name                                       NaN        NaN     NaN   
month_start_date                                   NaN        NaN     NaN   
year                                       2020.295381   0.456296  2020.0   
...                                                ...        ...     ...   
people_fully_vaccinated_per_100_month_end     7.013785  11.931419     0.0   
avg_government_response_stringency           57.611377  21.219293     0.0   
max_government_response_stringency           62.779968  22.142894     0.0   
min_government_response_stringency           52.140547  22.951128     0.0   
days_in_month                                29.022556   5.323341     1.0   

                                              25%        50%       75%     max  
country_code                                  NaN        NaN       NaN     NaN  
continent_name                                NaN        NaN       NaN     NaN  
country_name                                  NaN        NaN       NaN     NaN  
month_start_date                              NaN        NaN       NaN     NaN  
year                                       2020.0     2020.0    2021.0  2021.0  
...                                           ...        ...       ...     ...  
people_fully_vaccinated_per_100_month_end    0.67       2.61      8.11   97.47  
avg_government_response_stringency         44.382  59.812903  73.55394   100.0  
max_government_response_stringency          49.54      66.67     79.63   100.0  
min_government_response_stringency          36.34      54.63     70.37   100.0  
days_in_month                                30.0       31.0      31.0    31.0  

[67 rows x 11 columns]

In [13]:
final_covid.duplicated().sum() #no duplicates

np.int64(0)

**Fixing Null Vlaues Optimally**

In [14]:
final_covid.isnull().sum()

country_code                                    0
continent_name                                  0
country_name                                    0
month_start_date                                0
year                                            0
                                             ... 
people_fully_vaccinated_per_100_month_end    2468
avg_government_response_stringency            270
max_government_response_stringency            270
min_government_response_stringency            270
days_in_month                                   0
Length: 67, dtype: int64

In [15]:
final_covid.shape

(2793, 67)

**New Data Summary**
- Previouse data was too large and fetching the data on the daily basis about 82000 rows.
- Now we are fetching the data on the montly basis and now it's only about to 2793 rows and 67 columns

## **DATA CLEANING**

**1. Dealing with Missing Data**

**Initial Assessment**
- Perform a quick diagnostic of dataset completeness and coverage.

Summary Points:

- Prints dataset shape, date range, and number of countries.

- Calculates percentage of missing values per column.

- Identifies columns with over 20% and 50% missing data.

- Returns dataset and missing value summary for further cleaning.

In [16]:
# Initial assessment
print(f" Dataset Shape: {final_covid.shape}")
print(f" Date Range: {final_covid['month_start_date'].min()} to {final_covid['month_start_date'].max()}")
print(f" Countries: {final_covid['country_name'].nunique()}")
    
    # Missing value summary
missing_summary = (final_covid.isnull().sum() / len(final_covid) * 100).sort_values(ascending=False)
print(f" Columns with >50% missing: {(missing_summary > 50).sum()}")
print(f" Columns with >20% missing: {(missing_summary > 20).sum()}")

 Dataset Shape: (2793, 67)
 Date Range: 2020-01-01 to 2021-04-01
 Countries: 210
 Columns with >50% missing: 19
 Columns with >20% missing: 31


**clean_missing_values_by_priority()**

Purpose: Handle missing values strategically based on business importance.

## Summary Points:

- Converts month_start_date to datetime for time-based operations.

- Core Metrics (Population, GDP, HDI, Cases, Deaths):

- Demographics → Forward fill + median.

- COVID metrics → Replace early pandemic NaNs (before March 2020) with 0 and forward fill later values.

## Vaccination Data:

- Pre-2021 values replaced with 0 (pre-vaccine era).

- Post-2021 values interpolated per country.

- Testing & Policy Data:

- Forward filled by country.

-Default assumptions applied (e.g., 10% positivity rate, median stringency).

- Healthcare System Columns:

- Missing data flagged with new _available columns.

- Missing values replaced with 0 for consistency.

- Returns a clean dataset ready for validation.

In [17]:
def clean_missing_values_by_priority(final_covid):
    """Strategic cleaning based on business importance"""
   
    print(" Strategic Missing Value Cleaning...")
    final_covid_clean = final_covid.copy()
    final_covid_clean['month_start_date'] = pd.to_datetime(final_covid_clean['month_start_date'])
   
    #  Core Business Metrics (Demographics + COVID) 
    core_cols = [
        'total_population', 'human_development_index_score', 'gdp_per_capita_usd',
        'total_confirmed_cases_month_end', 'monthly_new_cases',
        'total_confirmed_deaths_month_end', 'monthly_new_deaths'
    ]
   
    for col in core_cols:
        if col in final_covid_clean.columns:
            if 'population' in col or 'gdp' in col or 'human_development' in col:
                # Demographics
                final_covid_clean[col] = final_covid_clean.groupby('country_name')[col].ffill()
                final_covid_clean[col] = final_covid_clean[col].fillna(final_covid_clean[col].median())
            else:
                #  early pandemic = 0, then forward fill
                early_mask = final_covid_clean['month_start_date'] < '2020-03-01'
                final_covid_clean.loc[early_mask, col] = final_covid_clean.loc[early_mask, col].fillna(0)
                final_covid_clean[col] = final_covid_clean.groupby('country_name')[col].ffill()
                final_covid_clean[col] = final_covid_clean[col].fillna(0)
   
    #  Vaccination Data (Timeline-Aware) 
    vaccination_cols = [col for col in final_covid_clean.columns if 'vaccin' in col or 'vaccine' in col]
   
    for col in vaccination_cols:
        # Pre-vaccine era = 0
        pre_vaccine = final_covid_clean['month_start_date'] < '2021-01-01'
        final_covid_clean.loc[pre_vaccine, col] = 0
       
        # Vaccine era
        vaccine_era = final_covid_clean['month_start_date'] >= '2021-01-01'
        if vaccine_era.sum() > 0:
            #  Using transform with interpolate for proper groupby application
            final_covid_clean.loc[vaccine_era, col] = (
                final_covid_clean.loc[vaccine_era]
                .groupby('country_name')[col]
                .transform(lambda x: x.interpolate(method='linear', limit_direction='both'))
            )
       
        final_covid_clean[col] = final_covid_clean[col].fillna(0)
   
    # Testing & Policy Data 
    testing_policy_cols = [col for col in final_covid_clean.columns if 'test' in col or 'stringency' in col]
   
    for col in testing_policy_cols:
        final_covid_clean[col] = final_covid_clean.groupby('country_name')[col].ffill()
       
        if 'positivity' in col:
            final_covid_clean[col] = final_covid_clean[col].fillna(10.0)  # Conservative 10%
        elif 'stringency' in col:
            final_covid_clean[col] = final_covid_clean[col].fillna(final_covid_clean[col].median())
            final_covid_clean[col] = final_covid_clean[col].clip(0, 100)
        else:
            final_covid_clean[col] = final_covid_clean[col].fillna(0)
   
    # Healthcare System (High Missing - Flag Approach) 
    healthcare_cols = [col for col in final_covid_clean.columns if 'icu' in col or 'hosp' in col]
   
    for col in healthcare_cols:
        # Create availability flag for analysis
        final_covid_clean[f'{col}_available'] = (~final_covid_clean[col].isnull()).astype(int)
        final_covid_clean[col] = final_covid_clean[col].fillna(0)
   
    print(" Strategic cleaning completed!")
    return final_covid_clean

In [18]:
final_covid = clean_missing_values_by_priority(final_covid)

 Strategic Missing Value Cleaning...
 Strategic cleaning completed!


In [19]:
final_covid.isnull().sum() 

country_code                                    0
continent_name                                  0
country_name                                    0
month_start_date                                0
year                                            0
                                               ..
peak_hospital_patients_available                0
avg_hospital_patients_per_million_available     0
peak_hospital_patients_per_million_available    0
total_monthly_icu_admissions_available          0
total_monthly_hospital_admissions_available     0
Length: 78, dtype: int64

### **Step 3: Business Logic Validation**
The business logic validation step involves checking whether the input data meets the requirements of the business rules. In this case, we will check whether the input data contains any missing values or outliers. If there are any issues, we will return an error message indicating what needs to be fixed.

In [20]:
def validate_business_logic(final_covid):
    """Validate and fix business rule violations"""
    
    print("🔍 Business Logic Validation...")
    final_covid_valid = final_covid.copy()
    fixes = 0
    
    #  Deaths ≤ Cases
    if all(c in final_covid_valid.columns for c in ['total_confirmed_deaths_month_end', 'total_confirmed_cases_month_end']):
        deaths_exceed = final_covid_valid['total_confirmed_deaths_month_end'] > final_covid_valid['total_confirmed_cases_month_end']
        if deaths_exceed.sum() > 0:
            final_covid_valid.loc[deaths_exceed, 'total_confirmed_deaths_month_end'] = \
                final_covid_valid.loc[deaths_exceed, 'total_confirmed_cases_month_end']
            fixes += deaths_exceed.sum()
            print(f"   ⚠️ Fixed {deaths_exceed.sum()} cases where deaths > cases")
    
    #  Percentages 0-100%
    percentage_cols = [col for col in final_covid_valid.columns if 
                      any(x in col for x in ['percent', 'per_100', 'positivity', 'fatality_rate'])]
    
    for col in percentage_cols:
        invalid = (final_covid_valid[col] > 100) | (final_covid_valid[col] < 0)
        if invalid.sum() > 0:
            final_covid_valid[col] = final_covid_valid[col].clip(0, 100)
            fixes += invalid.sum()
            print(f"   ⚠️ Fixed {invalid.sum()} invalid percentages in {col}")
    
    #  No negative counts
    count_cols = [col for col in final_covid_valid.columns if 
                 any(x in col for x in ['total_', 'monthly_', 'avg_', 'peak_'])]
    
    for col in count_cols:
        if final_covid_valid[col].dtype in ['int64', 'float64', 'Int64', 'float32']:
            negative = final_covid_valid[col] < 0
            if negative.sum() > 0:
                final_covid_valid.loc[negative, col] = 0
                fixes += negative.sum()
    
    print(f" Validation completed! Fixed {fixes} violations")
    return final_covid_valid


## Summary Points:

Rule 1: Fix cases where deaths > cases by setting deaths = cases.

Rule 2: Constrain percentage-based columns (e.g., positivity, fatality) within 0–100%.

Rule 3: Replace all negative counts with 0 (ensures no impossible negative data).

Tracks and reports total number of fixes applied.

Returns a logically consistent dataset.

In [21]:
final_covid = validate_business_logic(final_covid)

🔍 Business Logic Validation...
   ⚠️ Fixed 9 invalid percentages in vaccine_doses_per_100_month_end
   ⚠️ Fixed 1 invalid percentages in people_vaccinated_per_100_month_end
 Validation completed! Fixed 28 violations


### **Step 4: Type-Casting**


In [22]:

print(final_covid.dtypes)

country_code                                            object
continent_name                                          object
country_name                                            object
month_start_date                                datetime64[ns]
year                                                     int64
                                                     ...      
peak_hospital_patients_available                         int64
avg_hospital_patients_per_million_available              int64
peak_hospital_patients_per_million_available             int64
total_monthly_icu_admissions_available                   int64
total_monthly_hospital_admissions_available              int64
Length: 78, dtype: object


- No significant changes in Data types as they were already in the correct format

### **Step 5: Monthly Feature Engineering**


In [23]:
def create_monthly_business_features(final_covid):
    """Create features aligned with business objectives"""
    
    print(" Monthly Feature Engineering...")
    final_covid_enhanced = final_covid.copy()
    final_covid_enhanced = final_covid_enhanced.sort_values(['country_name', 'month_start_date'])
    
    # TEMPORAL FEATURES 
    final_covid_enhanced['quarter'] = final_covid_enhanced['month_start_date'].dt.quarter
    final_covid_enhanced['pandemic_year'] = final_covid_enhanced['year'] - 2020
    
    # Pandemic phases for temporal analysis
    final_covid_enhanced['pandemic_phase'] = pd.cut(
        final_covid_enhanced['pandemic_year'], 
        bins=[-1, 1, 2, 3, 10],
        labels=['Year_1', 'Year_2', 'Year_3', 'Later']
    )
    
    #  GROWTH INDICATORS 
    # Cases and vaccination growth rates
    growth_cols = ['total_confirmed_cases_month_end', 'people_vaccinated_per_100_month_end']
    for col in growth_cols:
        if col in final_covid_enhanced.columns:
            final_covid_enhanced[f'{col}_growth_rate'] = final_covid_enhanced.groupby('country_name')[col].pct_change()
    
    # POLICY EFFECTIVENESS 
    if all(c in final_covid_enhanced.columns for c in ['monthly_new_cases', 'avg_government_response_stringency']):
        final_covid_enhanced['policy_effectiveness'] = \
            final_covid_enhanced['monthly_new_cases'] / (final_covid_enhanced['avg_government_response_stringency'] + 1)
    
    # Policy consistency (within-month variation)
    if all(c in final_covid_enhanced.columns for c in ['max_government_response_stringency', 'min_government_response_stringency']):
        final_covid_enhanced['policy_consistency'] = \
            final_covid_enhanced['max_government_response_stringency'] - final_covid_enhanced['min_government_response_stringency']
    
    #  COMPARATIVE CATEGORIES 
    # Development levels for geographic comparison
    if 'human_development_index_score' in final_covid_enhanced.columns:
        final_covid_enhanced['development_level'] = pd.cut(
            final_covid_enhanced['human_development_index_score'],
            bins=[0, 0.55, 0.70, 0.80, 1.0],
            labels=['Low', 'Medium', 'High', 'Very_High']
        )
    
    # Vaccination progress levels
    if 'people_vaccinated_per_100_month_end' in final_covid_enhanced.columns:
        final_covid_enhanced['vaccination_level'] = pd.cut(
            final_covid_enhanced['people_vaccinated_per_100_month_end'],
            bins=[0, 30, 60, 80, 100],
            labels=['Low', 'Medium', 'High', 'Very_High'],
            include_lowest=True
        )
    
    print(" Feature engineering completed!")
    return final_covid_enhanced

In [24]:
final_covid =  create_monthly_business_features(final_covid)

 Monthly Feature Engineering...
 Feature engineering completed!


In [25]:
final_covid.shape

(2793, 87)

## Generate analytical features for trend and performance analysis.


- Adds temporal features: quarter, pandemic year, and phase (Year_1, Year_2, etc.).

- Calculates growth rates for cases and vaccination progress.

- Computes policy effectiveness (new cases ÷ stringency) and policy consistency (max–min stringency gap).

- Categorizes countries by development level (HDI-based) and vaccination level (coverage-based).

- Enhances dataset for business, epidemiological, and policy insights.

- Returns feature-enriched dataset.

## **Outlier Handling Techniques**

In [26]:

def clean_critical_data_quality_outliers(df):
    
  
    
    print(" CLEANING CRITICAL OUTLIERS (Data Quality Issues Only)")
    print("=" * 60)
    
    df_clean = df.copy()
    total_corrections = 0
    
    #  Deaths cannot exceed cases
    if 'total_confirmed_deaths' in df_clean.columns and 'total_confirmed_cases' in df_clean.columns:
        deaths_exceed = df_clean['total_confirmed_deaths'] > df_clean['total_confirmed_cases']
        count = deaths_exceed.sum()
        if count > 0:
            print(f"   ⚠️  Fixing {count} records where deaths > cases")
            df_clean.loc[deaths_exceed, 'total_confirmed_deaths'] = df_clean.loc[deaths_exceed, 'total_confirmed_cases']
            total_corrections += count
    
    #  Negative counts (impossible)
    negative_cols = ['total_confirmed_cases', 'total_confirmed_deaths', 'daily_new_cases', 'daily_new_deaths']
    
    for col in negative_cols:
        if col in df_clean.columns:
            negative_mask = df_clean[col] < 0
            count = negative_mask.sum()
            if count > 0:
                print(f"   ⚠️  Fixing {count} negative values in {col}")
                df_clean.loc[negative_mask, col] = 0
                total_corrections += count
    
    # Test positivity > 100% (impossible)
    if 'test_positivity_rate' in df_clean.columns:
        invalid_positivity = df_clean['test_positivity_rate'] > 100
        count = invalid_positivity.sum()
        if count > 0:
            print(f"     Fixing {count} test positivity rates > 100%")
            df_clean.loc[invalid_positivity, 'test_positivity_rate'] = 100
            total_corrections += count
    
    #  Invalid percentages
    percentage_cols = ['percent_aged_65_plus', 'percent_aged_70_plus', 'people_vaccinated_per_100_pop']
    
    for col in percentage_cols:
        if col in df_clean.columns:
            # Fix > 100%
            over_100 = df_clean[col] > 100
            count_over = over_100.sum()
            if count_over > 0:
                print(f"   Fixing {count_over} values > 100% in {col}")
                df_clean.loc[over_100, col] = 100
                total_corrections += count_over
            
            # Fix < 0%
            under_0 = df_clean[col] < 0
            count_under = under_0.sum()
            if count_under > 0:
                print(f"     Fixing {count_under} negative values in {col}")
                df_clean.loc[under_0, col] = 0
                total_corrections += count_under
    
    # Summary
    print(f"\n CRITICAL CLEANING COMPLETE!")
    if total_corrections > 0:
        print(f" Total corrections: {total_corrections}")
        print(f" Data ready for your existing pipeline")
    else:
        print(f" No critical issues found - excellent data quality!")
    print(f" Later: Epidemiological outlier analysis in EDA stage")
    
    return df_clean


In [27]:
final_covid = clean_critical_data_quality_outliers(final_covid)

 CLEANING CRITICAL OUTLIERS (Data Quality Issues Only)

 CRITICAL CLEANING COMPLETE!
 No critical issues found - excellent data quality!
 Later: Epidemiological outlier analysis in EDA stage


## Perfect now we push this cleaned data into our database and then we apply the EDA Later

In [28]:
final_covid.shape

(2793, 87)

In [29]:
colll= final_covid.info()
colll

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 87 columns):
 #   Column                                           Non-Null Count  Dtype         
---  ------                                           --------------  -----         
 0   country_code                                     2793 non-null   object        
 1   continent_name                                   2793 non-null   object        
 2   country_name                                     2793 non-null   object        
 3   month_start_date                                 2793 non-null   datetime64[ns]
 4   year                                             2793 non-null   int64         
 5   month                                            2793 non-null   int64         
 6   total_population                                 2793 non-null   float64       
 7   population_per_sq_km                             2712 non-null   float64       
 8   median_age_years                      

In [30]:
cursor.execute("""CREATE TABLE IF NOT EXISTS final_covid(
`id` INT NOT NULL AUTO_INCREMENT,
`country_code` VARCHAR(255),
`continent_name` VARCHAR(255),
`country_name` VARCHAR(255),
`month_start_date` DATE,
`year` BIGINT,
`month` BIGINT,
`total_population` DOUBLE,
`population_per_sq_km` DOUBLE,
`median_age_years` DOUBLE,
`percent_aged_65_plus` DOUBLE,
`percent_aged_70_plus` DOUBLE,
`gdp_per_capita_usd` DOUBLE,
`percent_in_extreme_poverty` DOUBLE,
`cardiovascular_death_rate` DOUBLE,
`diabetes_prevalence_percent` DOUBLE,
`female_smoking_rate` DOUBLE,
`male_smoking_rate` DOUBLE,
`handwashing_facilities_percent` DOUBLE,
`hospital_beds_per_1000_pop` DOUBLE,
`life_expectancy_years` DOUBLE,
`human_development_index_score` DOUBLE,
`total_confirmed_cases_month_end` DOUBLE,
`monthly_new_cases` DOUBLE,
`avg_daily_cases_7day` DOUBLE,
`cases_per_million_month_end` DOUBLE,
`avg_daily_cases_per_million` DOUBLE,
`avg_daily_cases_per_million_7day` DOUBLE,
`total_confirmed_deaths_month_end` DOUBLE,
`monthly_new_deaths` DOUBLE,
`avg_daily_deaths_7day` DOUBLE,
`deaths_per_million_month_end` DOUBLE,
`avg_daily_deaths_per_million` DOUBLE,
`avg_daily_deaths_per_million_7day` DOUBLE,
`case_fatality_rate_percent` DOUBLE,
`avg_virus_reproduction_rate` DOUBLE,
`max_virus_reproduction_rate` DOUBLE,
`avg_icu_patients` DOUBLE,
`peak_icu_patients` DOUBLE,
`avg_icu_patients_per_million` DOUBLE,
`peak_icu_patients_per_million` DOUBLE,
`avg_hospital_patients` DOUBLE,
`peak_hospital_patients` DOUBLE,
`avg_hospital_patients_per_million` DOUBLE,
`peak_hospital_patients_per_million` DOUBLE,
`total_monthly_icu_admissions` DOUBLE,
`total_monthly_hospital_admissions` DOUBLE,
`total_monthly_tests_conducted` DOUBLE,
`total_tests_conducted_month_end` DOUBLE,
`tests_per_thousand_month_end` DOUBLE,
`avg_daily_tests_per_thousand` DOUBLE,
`avg_daily_tests_7day` DOUBLE,
`avg_test_positivity_rate` DOUBLE,
`min_test_positivity_rate` DOUBLE,
`max_test_positivity_rate` DOUBLE,
`avg_tests_required_per_case` DOUBLE,
`total_vaccine_doses_month_end` DOUBLE,
`people_with_at_least_one_dose_month_end` DOUBLE,
`people_fully_vaccinated_month_end` DOUBLE,
`monthly_vaccinations` DOUBLE,
`avg_daily_vaccinations_7day` DOUBLE,
`vaccine_doses_per_100_month_end` DOUBLE,
`people_vaccinated_per_100_month_end` DOUBLE,
`people_fully_vaccinated_per_100_month_end` DOUBLE,
`avg_government_response_stringency` DOUBLE,
`max_government_response_stringency` DOUBLE,
`min_government_response_stringency` DOUBLE,
`days_in_month` BIGINT,
`hospital_beds_per_1000_pop_available` BIGINT,
`avg_icu_patients_available` BIGINT,
`peak_icu_patients_available` BIGINT,
`avg_icu_patients_per_million_available` BIGINT,
`peak_icu_patients_per_million_available` BIGINT,
`avg_hospital_patients_available` BIGINT,
`peak_hospital_patients_available` BIGINT,
`avg_hospital_patients_per_million_available` BIGINT,
`peak_hospital_patients_per_million_available` BIGINT,
`total_monthly_icu_admissions_available` BIGINT,
`total_monthly_hospital_admissions_available` BIGINT,
`quarter` INT,
`pandemic_year` BIGINT,
`pandemic_phase` VARCHAR(100),
`total_confirmed_cases_month_end_growth_rate` DOUBLE,
`people_vaccinated_per_100_month_end_growth_rate` DOUBLE,
`policy_effectiveness` DOUBLE,
`policy_consistency` DOUBLE,
`development_level` VARCHAR(100),
`vaccination_level` VARCHAR(100),
PRIMARY KEY (`id`),
UNIQUE KEY `uq_country_month` (`country_code`, `month_start_date`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;"""

)

conn.commit()
cursor.close()
conn.close()

In [31]:
engine = create_engine("mysql+mysqlconnector://root:ibadat@localhost/covid_db")

# Key operations:
final_covid_clean = final_covid.replace([np.inf, -np.inf], None)
final_covid_clean.to_sql('final_covid', con=engine, if_exists='replace', 
                        index=False, chunksize=1000)

2793

In [32]:
final_covid_clean  # Now this is our new cleaned data

country_code continent_name country_name month_start_date  year  month  \
0             AFG           Asia  Afghanistan       2020-02-01  2020      2   
1             AFG           Asia  Afghanistan       2020-03-01  2020      3   
2             AFG           Asia  Afghanistan       2020-04-01  2020      4   
3             AFG           Asia  Afghanistan       2020-05-01  2020      5   
4             AFG           Asia  Afghanistan       2020-06-01  2020      6   
...           ...            ...          ...              ...   ...    ...   
2788          ZWE         Africa     Zimbabwe       2020-12-01  2020     12   
2789          ZWE         Africa     Zimbabwe       2021-01-01  2021      1   
2790          ZWE         Africa     Zimbabwe       2021-02-01  2021      2   
2791          ZWE         Africa     Zimbabwe       2021-03-01  2021      3   
2792          ZWE         Africa     Zimbabwe       2021-04-01  2021      4   

     total_population population_per_sq_km median_age_years  \
0          38928341.0               54.422             18.6   
1          38928341.0               54.422             18.6   
2          38928341.0               54.422             18.6   
3          38928341.0               54.422             18.6   
4          38928341.0               54.422             18.6   
...               ...                  ...              ...   
2788       14862927.0               42.729             19.6   
2789       14862927.0               42.729             19.6   
2790       14862927.0               42.729             19.6   
2791       14862927.0               42.729             19.6   
2792       14862927.0               42.729             19.6   

     percent_aged_65_plus  ... total_monthly_hospital_admissions_available  \
0                   2.581  ...                                           0   
1                   2.581  ...                                           0   
2                   2.581  ...                                           0   
3                   2.581  ...                                           0   
4                   2.581  ...                                           0   
...                   ...  ...                                         ...   
2788                2.822  ...                                           0   
2789                2.822  ...                                           0   
2790                2.822  ...                                           0   
2791                2.822  ...                                           0   
2792                2.822  ...                                           0   

     quarter pandemic_year pandemic_phase  \
0          1             0         Year_1   
1          1             0         Year_1   
2          2             0         Year_1   
3          2             0         Year_1   
4          2             0         Year_1   
...      ...           ...            ...   
2788       4             0         Year_1   
2789       1             1         Year_1   
2790       1             1         Year_1   
2791       1             1         Year_1   
2792       2             1         Year_1   

     total_confirmed_cases_month_end_growth_rate  \
0                                            NaN   
1                                          174.0   
2                                      11.154286   
3                                       6.149976   
4                                       1.071739   
...                                          ...   
2788                                    0.393668   
2789                                    1.407731   
2790                                    0.080897   
2791                                    0.021973   
2792                                    0.037281   

     people_vaccinated_per_100_month_end_growth_rate policy_effectiveness  \
0                                                NaN             0.107181   
1                                         